File: task1/.js
function minWindow(s, t) {
  if (!s || !t || s.length < t.length) {
    return "";
  }
  const tCharCounts = new Map(); // Счетчик символов для строки T
  const windowCharCounts = new Map(); // Счетчик символов для текущего окна
  for (const char of t) {
    tCharCounts.set(char, (tCharCounts.get(char) || 0) + 1);
  }
  let left = 0;
  let right = 0;
  let minLen = Infinity;
  let minStart = 0;
  let matched = 0; // Количество символов из T, найденных в текущем окне
  while (right < s.length) {
    const char = s[right];
    if (tCharCounts.has(char)) {
      windowCharCounts.set(char, (windowCharCounts.get(char) || 0) + 1);
      if (windowCharCounts.get(char) === tCharCounts.get(char)) {
        matched++; // Увеличиваем счетчик, если количество символов совпало
      }
    }
    while (matched === tCharCounts.size) { // Когда все символы T найдены в окне
      if (right - left + 1 < minLen) {
        minLen = right - left + 1;
        minStart = left;
      }
      const leftChar = s[left];
      if (tCharCounts.has(leftChar)) {
        windowCharCounts.set(leftChar, windowCharCounts.get(leftChar) - 1);
        if (windowCharCounts.get(leftChar) < tCharCounts.get(leftChar)) {
          matched--; // Уменьшаем счетчик, так как символ больше не удовлетворяет условию
        }
      }
      left++;
    }
    right++;
  }
  return minLen === Infinity ? "" : s.substring(minStart, minStart + minLen);
}
const s = "ADOBECODEBANC";
const t = "ABC";
const result = minWindow(s, t);
console.log(result);

File: task1/.swift
func minWindow(s: String, t: String) -> String {
    guard !s.isEmpty && !t.isEmpty && s.count >= t.count else { return "" }

    var tCharCounts: [Character: Int] = [:] // Счетчик символов для строки T
    var windowCharCounts: [Character: Int] = [:] // Счетчик символов для текущего окна

    for char in t {
        tCharCounts[char, default: 0] += 1
    }

    var left = 0
    var right = 0
    var minLen = Int.max
    var minStart = 0
    var matched = 0 // Количество символов из T, найденных в текущем окне

    let sArray = Array(s) // Преобразуем строку S в массив символов для удобства доступа

    while right < sArray.count {
        let char = sArray[right]

        if tCharCounts[char] != nil {
            windowCharCounts[char, default: 0] += 1
            if windowCharCounts[char] == tCharCounts[char] {
                matched += 1 // Увеличиваем счетчик, если количество символов совпало
            }
        }

        while matched == tCharCounts.count { // Когда все символы T найдены в окне
            if right - left + 1 < minLen {
                minLen = right - left + 1
                minStart = left
            }

            let leftChar = sArray[left]
            if tCharCounts[leftChar] != nil {
                windowCharCounts[leftChar, default: 0] -= 1
                if windowCharCounts[leftChar, default: 0] < tCharCounts[leftChar, default: 0] {
                    matched -= 1 // Уменьшаем счетчик, так как символ больше не удовлетворяет условию
                }
            }
            left += 1
        }
        right += 1
    }

    return minLen == Int.max ? "" : String(sArray[minStart..<minStart+minLen])
}

let s = "ADOBECODEBANC"
let t = "ABC"
let result = minWindow(s: s, t: t)
print(result) 
